extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js

```diff
@@ -0,0 +1,618 @@
+      rootBoundEnd - boundingRectStart : boundingRectEnd;
+    const percInView = lengthInView / (boundingRectEnd - boundingRectStart);
+    return Math.max(0, Math.min(1, percInView)) || 0;
+  }
+
+  /**
+   * Handles serializing and sending messages to the safeframe.
+   * @param {!Object} payload
+   * @param {string} serviceName
+   * @private
+   */
+  sendMessage_(payload, serviceName) {
+    dev().assert(this.iframe_.contentWindow,
+        'Frame contentWindow unavailable.');
+    const message = dict();
+    message[MESSAGE_FIELDS.CHANNEL] = this.channel;
+    message[MESSAGE_FIELDS.PAYLOAD] = JSON.stringify(
+        /** @type {!JsonObject} */(payload));
+    message[MESSAGE_FIELDS.SERVICE] = serviceName;
+    message[MESSAGE_FIELDS.SENTINEL] = this.sentinel_;
+    message[MESSAGE_FIELDS.ENDPOINT_IDENTITY] = this.endpointIdentity_;
+    this.iframe_.contentWindow./*OK*/postMessage(
+        JSON.stringify(message), SAFEFRAME_ORIGIN);
+  }
+
+  /**
+   * Routes messages to their appropriate handler.
+   * @param {!JsonObject} payload
+   * @param {string} service
+   */
+  processMessage(payload, service) {
+    // We are not logging unexpected messages, and some expected
+    // messages are being dropped, like init_done, as we don't need them.
+    switch (service) {
+      case SERVICE.CREATIVE_GEOMETRY_UPDATE:
+        this.handleFluidMessage_(payload);
+        break;
+      case SERVICE.EXPAND_REQUEST:
+        this.handleExpandRequest_(payload);
+        break;
+      case SERVICE.REGISTER_DONE:
+        this.isRegistered_ = true;
+        break;
+      case SERVICE.COLLAPSE_REQUEST:
+        this.handleCollapseRequest_();
+        break;
+      default:
+        break;
+    }
+  }
+
+  /**
+   * @param {!JsonObject} payload
+   * @private
+   */
+  handleExpandRequest_(payload) {
+    if (!this.isCollapsed_ || !this.isRegistered_) {
+      return;
+    }
+    const expandHeight = Number(this.iframe_.height) +
+          payload['expand_b'] + payload['expand_t'];
+    const expandWidth = Number(this.iframe_.width) +
+          payload['expand_r'] + payload['expand_l'];
+    // Verify that if expanding by push, that expandByPush is allowed.
+    // If expanding by overlay, verify that expandByOverlay is allowed,
+    // and that we are only expanding within the bounds of the amp-ad.
+    if ((payload['push'] && !this.expandByPush_) ||
+        (!payload['push'] && !this.expandByOverlay_ &&
+         (expandWidth > this.creativeSize_.width ||
+          expandHeight > this.creativeSize_.height))) {
+      return;
+    }
+    // Can't expand to greater than the viewport size
+    if (expandHeight > this.viewport_.getSize().height ||
+        expandWidth > this.viewport_.getSize().width) {
+      return;
+    }
+    this.handleSizeChange(expandHeight,
+        expandWidth,
+        SERVICE.EXPAND_RESPONSE);
+  }
+
+  /**
+   * @private
+   */
+  handleCollapseRequest_() {
+    if (this.isCollapsed_ || !this.isRegistered_) {
```

Comment:
We want to check that it was previously expanded right?  Does this do that?

Reply:
yes, this is saying "if it is already collapsed, don't collapse", i.e. it only collapses if expanded.